### PR TITLE
Move DB functionality to service layer – OrderingPartyController

### PR DESCRIPTION
**Pull request by RahulJha-mastek**
Story [AB#13236](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/13236)
Task [AB#13835](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/13835)

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderingPartyController.cs [hunk 1]
```diff
@@ -55,25 +55,23 @@ public async Task<ActionResult> UpdateAsync(CallOffId callOffId, OrderingPartyMo
...
 
-            var orderingParty = order.OrderingParty;
-            orderingParty.Name = model.Name;
-            orderingParty.OdsCode = model.OdsCode;
+            OrderingParty orderingParty = new OrderingParty
+                {
+                    Name = model.Name,
+                    OdsCode = model.OdsCode,
+                    Address = contactDetailsService.AddOrUpdateAddress(order.OrderingParty.Address, model.Address),
+                };
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderingPartyController.cs:68]: We can use `var` here. Also, can you reduce the indent of the braces by one tab? (Just to keep the styling consistent with other places in the code.)

File: src/NHSD.BuyingCatalogue.Ordering.Api/Services/ContactDetailsService.cs [hunk 2]
```diff
@@ -1,26 +1,29 @@
...
+                existingAddress.Line1 = newOrUpdatedAddress.Line1;
+                existingAddress.Line2 = newOrUpdatedAddress.Line2;
+                existingAddress.Line3 = newOrUpdatedAddress.Line3;
+                existingAddress.Line4 = newOrUpdatedAddress.Line4;
+                existingAddress.Line5 = newOrUpdatedAddress.Line5;
+                existingAddress.Town = newOrUpdatedAddress.Town;
+                existingAddress.County = newOrUpdatedAddress.County;
+                existingAddress.Postcode = newOrUpdatedAddress.Postcode;
+                existingAddress.Country = newOrUpdatedAddress.Country;
+            }
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Services/ContactDetailsService.cs:25]: Let's invert the `if` here (to reduce the amount of nested code).

File: src/NHSD.BuyingCatalogue.Ordering.Api/Services/ContactDetailsService.cs [hunk 3]
```diff
@@ -29,9 +32,13 @@ public Contact AddOrUpdatePrimaryContact(Contact existingContact, PrimaryContact
...
-            context.Entry(existingContact).CurrentValues.SetValues(newOrUpdatedContact);
-            existingContact.Email = newOrUpdatedContact.EmailAddress;
-            existingContact.Phone = newOrUpdatedContact.TelephoneNumber;
+            if (newOrUpdatedContact is not null)
+            {
+                existingContact.FirstName = newOrUpdatedContact.FirstName;
+                existingContact.LastName = newOrUpdatedContact.LastName;
+                existingContact.Email = newOrUpdatedContact.EmailAddress;
+                existingContact.Phone = newOrUpdatedContact.TelephoneNumber;
+            }
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Ordering.Api/Services/ContactDetailsService.cs:41]: As above, let's invert the `if` here (to reduce the amount of nested code).

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs [hunk 4]
```diff
@@ -49,20 +48,20 @@ public static void Constructor_VerifyGuardClauses()
...
             order.OrderingPartyContact.Should().NotBeNull();
 
-            context.Add(order);
-            await context.SaveChangesAsync();
+            service.Setup(o => o.GetOrder(callOffId)).ReturnsAsync(order);
 
             var expectedValue = new OrderingPartyModel(order.OrderingParty, order.OrderingPartyContact);
 
             var response = await controller.GetAsync(callOffId);
+            service.Verify(o => o.GetOrder(callOffId));
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs:64]: I don't think we need to `Verify` here as the test covers this as part of its assertion.

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs [hunk 5]
```diff
@@ -79,30 +78,39 @@ public static void Constructor_VerifyGuardClauses()
...
         public static async Task UpdateAsync_OrderDoesNotExist_ReturnsNotFound(
+            [Frozen] Mock<IOrderingPartyService> service,
             CallOffId callOffId,
             OrderingPartyModel model,
             OrderingPartyController controller)
         {
+            service.Setup(o => o.GetOrder(callOffId)).ReturnsAsync((Order)null);
+
             var response = await controller.UpdateAsync(callOffId, model);
+            service.Verify(o => o.GetOrder(callOffId));
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs:89]: As above, I don't think we need to `Verify` here as the test covers this as part of its assertion.

If you think it's worthwhile verifying somewhere that `GetAsync` invokes `GetOrder`, I'd recommend adding a separate test for that.

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs [hunk 6]
```diff
@@ -79,30 +78,39 @@ public static void Constructor_VerifyGuardClauses()
...
+            service.Setup(o => o.GetOrder(callOffId)).ReturnsAsync(order);
+            service.Setup(o => o.SetOrderingParty(order, It.IsAny<OrderingParty>(), It.IsAny<Contact>())).Callback(() =>
+            {
+                order.OrderingParty.Name = model.Name;
+                order.OrderingParty.OdsCode = model.OdsCode;
+            });
 
             await controller.UpdateAsync(callOffId, model);
+            service.Verify(o => o.GetOrder(callOffId));
+            service.Verify(o => o.SetOrderingParty(order, It.IsAny<OrderingParty>(), It.IsAny<Contact>()));
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs:113]: As above, I don't think we need to `Verify` tests here as the test covers this as part of its assertion.

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs [hunk 7]
```diff
@@ -113,78 +121,85 @@ public static void Constructor_VerifyGuardClauses()
...
+            {
+                order.OrderingParty.Name = model.Name;
+                order.OrderingParty.OdsCode = model.OdsCode;
+            });
+            contactDetailsService.Setup(o => o.AddOrUpdateAddress(It.IsAny<Address>(), It.IsAny<AddressModel>()))
+                .Returns(order.OrderingParty.Address);
 
+            await controller.UpdateAsync(callOffId, model);
+            service.Verify(o => o.GetOrder(callOffId));
+            service.Verify(o => o.SetOrderingParty(order, It.IsAny<OrderingParty>(), It.IsAny<Contact>()));
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs:143]: I don't think we need these additional `Verify` tests here as we're testing that `AddOrUpdateAddress` is invoked here.

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs [hunk 8]
```diff
@@ -113,78 +121,85 @@ public static void Constructor_VerifyGuardClauses()
...
+                order.OrderingParty.OdsCode = model.OdsCode;
+            });
+            contactDetailsService.Setup(o => o.AddOrUpdatePrimaryContact(
+                    It.Is<Contact>(c => c == originalContact),
+                    It.Is<PrimaryContactModel>(c => c == model.PrimaryContact)))
+                .Returns(order.OrderingPartyContact);
 
+            await controller.UpdateAsync(callOffId, model);
+            service.Verify(o => o.GetOrder(callOffId));
+            service.Verify(o => o.SetOrderingParty(order, It.IsAny<OrderingParty>(), It.IsAny<Contact>()));
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs:176]: I don't think we need these additional `Verify` tests here as we're testing that `AddOrUpdatePrimaryContact` is invoked here.

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs [hunk 9]
```diff
@@ -113,78 +121,85 @@ public static void Constructor_VerifyGuardClauses()
...
+            service.Setup(o => o.GetOrder(callOffId)).ReturnsAsync(order);
+            service.Setup(o => o.SetOrderingParty(order, It.IsAny<OrderingParty>(), It.IsAny<Contact>())).Callback(() =>
+            {
+                order.OrderingParty.Name = model.Name;
+                order.OrderingParty.OdsCode = model.OdsCode;
+            });
 
             var result = await controller.UpdateAsync(callOffId, model);
+            service.Verify(o => o.GetOrder(callOffId));
+            service.Verify(o => o.SetOrderingParty(order, It.IsAny<OrderingParty>(), It.IsAny<Contact>()));
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs:202]: I don't think we need these additional `Verify` tests here as these are covered as part of the test.

However, If you think it's worthwhile verifying somewhere that `UpdateAsync` invokes `GetOrder` and `SetOrderingParty`, I'd recommend adding two separate tests for that.

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Services/ContactDetailServiceTests.cs [hunk 10]
```diff
@@ -0,0 +1,166 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.Threading.Tasks;
+using AutoFixture.NUnit3;
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Services/ContactDetailServiceTests.cs:3]: Are the following `using` statements required?

```c#
using System.Threading.Tasks;
using AutoFixture.NUnit3;
```

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Services/ContactDetailServiceTests.cs [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+using NHSD.BuyingCatalogue.Ordering.Api.UnitTests.AutoFixture;
+using NHSD.BuyingCatalogue.Ordering.Domain;
+using NUnit.Framework;
+
+namespace NHSD.BuyingCatalogue.Ordering.Api.UnitTests.Services
+{
+    [TestFixture]
+    [Parallelizable(ParallelScope.All)]
+    [SuppressMessage("ReSharper", "NUnit.MethodWithParametersAndTestAttribute", Justification = "False positive")]
+    internal static class ContactDetailServiceTests
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Services/ContactDetailServiceTests.cs:16]: Should be `ContactDetailsServiceTests` to match the service name.

**Review by bjss-chris — CHANGES_REQUESTED** [review]
Looks good overall – just a few suggestions, mostly around the unit tests.

File: src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderingPartyController.cs (hunk 1 above)
Comment (RahulJha-mastek) [reply src/NHSD.BuyingCatalogue.Ordering.Api/Controllers/OrderingPartyController.cs]: Updated with `var` keyword and updated indentation

File: src/NHSD.BuyingCatalogue.Ordering.Api/Services/ContactDetailsService.cs (hunk 2 above)
Comment (RahulJha-mastek) [reply src/NHSD.BuyingCatalogue.Ordering.Api/Services/ContactDetailsService.cs]: Inverted `If` statement to check `Null`

File: src/NHSD.BuyingCatalogue.Ordering.Api/Services/ContactDetailsService.cs (hunk 3 above)
Comment (RahulJha-mastek) [reply src/NHSD.BuyingCatalogue.Ordering.Api/Services/ContactDetailsService.cs]: Inverted `If` statement to check `Null`

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs (hunk 4 above)
Comment (RahulJha-mastek) [reply tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs]: Removed `service.Verify()`

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs (hunk 5 above)
Comment (RahulJha-mastek) [reply tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs]: Removed `service.Verify()`
And added new Unit test method to test `GetOrder` Invoke

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs (hunk 6 above)
Comment (RahulJha-mastek) [reply tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs]: Removed `service.Verify()`

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs (hunk 7 above)
Comment (RahulJha-mastek) [reply tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs]: Removed `service.Verify()`

**Review by bjss-chris — APPROVED** [review]

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs (hunk 8 above)
Comment (RahulJha-mastek) [reply tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs]: Removed `service.Verify()`

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs (hunk 9 above)
Comment (RahulJha-mastek) [reply tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Controllers/OrderingPartyControllerTests.cs]: Removed `service.Verify()`
And added new test method for `SetOrderingParty` invoke

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Services/ContactDetailServiceTests.cs (hunk 10 above)
Comment (RahulJha-mastek) [reply tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Services/ContactDetailServiceTests.cs]: Those are unused. Removed those

File: tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Services/ContactDetailServiceTests.cs (hunk 11 above)
Comment (RahulJha-mastek) [reply tests/NHSD.BuyingCatalogue.Ordering.Api.UnitTests/Services/ContactDetailServiceTests.cs]: Updated service name to `ContactDetailsServiceTests`

**Pull request merged**